contracts/sources/chanllenge.move

```diff
@@ -0,0 +1,302 @@
+
+        let p2 = memory::read_memory(&mut mem, finalSystemState, 0x30000800);
+        assert!(p2 == 0x1337F00d, ERR_FINAL_STATE_ROOT_NOT_WRITTEN);
+
+        assert!(memory::read_memory(&mut mem, finalSystemState, 0x30000804) == vector_to_u64(hash_value::bytes(&assertionRoots)),
+            ERR_MIPS_MACHINE_DIFFERENT_STATE_ROOT);
+
+        let challenge_id = get_global_challenge_id(signer);
+
+        set_challenge_challenger(signer, challenge_id,
+            Signer::address_of(signer),
+            blockNumberN_hash,
+            copy start_state,
+            copy start_state,
+            step_count,
+            finalSystemState,
+            0,
+            step_count
+        );
+
+        memory::return_mem(mem);
+
+        0
+    }
+
+    public fun is_searching(challenge_id: u64): bool acquires Challenges {
+        let contract_signer = ContractAccount::get_contract_signer();
+        let challenges = borrow_global_mut<Challenges>(Signer::address_of(&contract_signer));
+        let c = Vector::borrow_mut(&mut challenges.value, challenge_id);
+
+        assert!(c.challenger != @0x0, ERR_INVALID_CHALLENGE);
+
+        (c.L + 1 != c.R)
+    }
+
+    public fun get_step_number(challenge_id: u64): u64 acquires Challenges {
+        let contract_signer = ContractAccount::get_contract_signer();
+        let challenges = borrow_global_mut<Challenges>(Signer::address_of(&contract_signer));
+        let c = Vector::borrow_mut(&mut challenges.value, challenge_id);
+
+        assert!(c.challenger != @0x0, ERR_INVALID_CHALLENGE);
+
+        (c.L + c.R) / 2
+    }
+
+    public fun get_proposed_state(challenge_id: u64): HashValue acquires Challenges {
+        let step_number = get_step_number(challenge_id);
+
+        let contract_signer = ContractAccount::get_contract_signer();
+        let challenges = borrow_global_mut<Challenges>(Signer::address_of(&contract_signer));
+        let c = Vector::borrow_mut(&mut challenges.value, challenge_id);
+
+        assert!(c.challenger != @0x0, ERR_INVALID_CHALLENGE);
+
+        let borrowed_state = Table::borrow(&c.asserted_state, step_number);
+        *borrowed_state
+    }
+
+    // the challenger call this function to submit the state hash
+    // for next step in the binary search
+    public fun propose_state(signer: &signer, challenge_id: u64, state_hash: HashValue) acquires Challenges {
+        assert!(is_searching(challenge_id), ERR_MUST_BE_SEARCHING);
+        let step_number = get_step_number(challenge_id);
+
+        let contract_signer = ContractAccount::get_contract_signer();
+        let challenges = borrow_global_mut<Challenges>(Signer::address_of(&contract_signer));
+        let c = Vector::borrow_mut(&mut challenges.value, challenge_id);
+
+        assert!(c.challenger != @0x0, ERR_INVALID_CHALLENGE);
+        assert!(c.challenger == Signer::address_of(signer), ERR_MUST_BE_CHALLENGER);
+
+        let state = Table::borrow(&c.asserted_state, step_number);
```

Comment:
Table::borrow will raise err if the key is not found

Reply:
problem has fixed: use a function to check key exists in table, if not then throw an error.

```move
fun borrow_table_value(table: &Table::Table<u64, HashValue>, key: u64): &HashValue {
    assert!(Table::contains(table, key), ERR_TABLE_KEY_NOT_EXISTS);
    Table::borrow(table, key)
}
```